proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2245 @@
+implementing linear variants of ``base`` functions remain compatible
+with ``base`` (e.g. there need not be two ``Maybe`` types, two list
+types etc).
+
+The only function which will need to change is ``($)`` because its
+typing rule is built in the type checker. Ignoring the details about
+levity and higher-rank polymorphism in the typing rule, the type
+``($)`` will be:
+
+::
+
+  ($) :: (a :p-> b) ⊸ a :p-> b
+
+Defining a linear variant of ``base`` is out of scope of this
+proposal. Possible future standardisation of the library content is
+the competence of the Core Libraries Committee (CLC). For expository
+purposes of the next sections, however, we assume that such a library
+will at least define the following data type:
+
+::
+
+   data Unrestricted a where
+     Unrestricted :: a -> Unrestricted a
+
+See the paper for intuitions about the ``Unrestricted`` data type.
+
+Multiplicities
+--------------
+
+.. _Multiplicities:
+
+So far, we have considered only two multiplicities: ``1`` and ``ω``.
+But the metatheory works with any so-called sup-semi-lattice-ordered
+semi-ring (without a 0) of multiplicities. That is: there is a 1,
+a sum and a product with the usual distributivity laws, a (computable)
+order compatible with the sum and product, such that each pair of
+multiplicities has a (computable) join. Even if there is only two
+multiplicities in this proposal, the proposal is structured to allow
+future extensions.
+
+Here is the definition of sum, product and order for this proposal's
+multiplicities (in Haskell pseudo-syntax):
+
+::
+
+   _ + _ = ω
+
+   1 * x = x
+   x * 1 = x
+   ω * ω = ω
+
+   _ ⩽ ω = True
+   x ⩽ y = x == y
+
+Every variable in the environment is annotated with its multiplicity,
+which constrains how it can be used. A variable *usage* is said to be
+of multiplicity ``p``, or ``0``, in a term ``u`` if:
+
+- ``p=0`` and ``x`` is not free in ``u``
+- ``p=1`` and ``u = x``
+- ``p=p1+q*p2`` and ``u = u1 u2`` with ``u1 :: a :q-> b`` and the
+  usage of ``x`` in ``u1`` is ``p1``, and in ``u2`` is ``p2``
+- ``u = λy. v`` and the usage of ``x`` in ``v`` is ``p``.
+
+A variable's usage is correct if it is smaller than or equal to the
+multiplicity annotation of the variable (note that 0 is *not* smaller
+than one). Incorrect usage results in a type error. This definition is
+close to the intended implementation of multiplicities. The `paper
+<https://arxiv.org/abs/1710.09756>`_ has a more declarative
+definition.
+
+The multiplicity of a variable introduced by a λ-abstraction is taken
+from the surrounding typing information (typically a type annotation
+on an equation). For instance
+
+::
+
+  foo :: A :p-> B
+  foo x = …  -- x has multiplicity p
+
+The above takes care of the pure λ-calculus part of Haskell. We also
+need to consider ``let`` and ``case``.
+
+A ``let`` binding is considered to have an implicit multiplicity
+annotation (the annotation is inferred). The variables introduced by a
```

Comment:
18. This section makes me believe that all bindings in one `let` have the same multiplicity.